torch/csrc/jit/script/compiler.cpp

```
@@ -106,14 +249,111 @@ struct to_ir {
+    for (auto runner = environment_stack->next; runner; runner = runner->next) {
+      if (runner->value_table.count(name)) {
+        return runner->value_table.at(name);
+      }
+    }
+    return nullptr;
+  }
+
+  Value* createBlockInput(const std::string& name) {
+    auto& curr_frame = *environment_stack;
+    Block* b = curr_frame.b;
+
+    // Create the input
+    Value* new_input = b->addInput();
+
+    // Associate this name with this value
+    curr_frame.value_table[name] = new_input;
+
+    // List as a positional input
+    curr_frame.positional_inputs.push_back(name);
+
+    return new_input;
+  }
+
   void setVar(const std::string& name, Value* value) {
-    value_table[name] = value;
+    auto& curr_frame = *environment_stack;
+
+    Symbol owning_kind = Symbol();
+    if (curr_frame.b) {
+      owning_kind = curr_frame.b->owningNode()->kind();
+    }
+
+    if (owning_kind == Symbol("Loop") || owning_kind == Symbol("If")) {
+      // Overwriting an existing value means it's already been
+      // accounted for.
+      if (findInThisFrame(name)) {
+        curr_frame.value_table[name] = value;
+        return;
+      }
+
+      // Value not here. search in parent scopes
+      if (findInParentFrame(name)) {
+        // Writing to a value in a parent frame. Make this a
+        // loop-carried dependency.
+        if (owning_kind == Symbol("Loop"))
+          createBlockInput(name);
+
+        // Overwrite in value map
+        curr_frame.value_table[name] = value;
+      } else {
+        // Not accessing a value in enclosing scope. Make a new
+        // value as usual
+        curr_frame.value_table[name] = value;
+      }
+    } else {
+      curr_frame.value_table[name] = value;
+    }
   }
 
   Value* getVar(const Ident& ident) {
-    if (value_table.count(ident.name()) == 0)
-      throw ErrorReport(ident) << "undefined value " << ident.name();
-    return value_table[ident.name()];
+    try {
+      return getVar(ident.name());
```

I think `getVar` just returns null when it doesn't find stuff. So you need to throw an exception in this case. I don't think this try..catch does anything meaningful.